presto-main/src/main/java/io/prestosql/server/QueryResource.java

```diff
@@ -68,55 +83,76 @@ public QueryResource(DispatchManager dispatchManager)
 
     @GET
     @Path("{queryId}")
-    public Response getQueryInfo(@PathParam("queryId") QueryId queryId)
+    public Response getQueryInfo(@PathParam("queryId") QueryId queryId, @Context HttpServletRequest servletRequest)
     {
         requireNonNull(queryId, "queryId is null");
 
         Optional<QueryInfo> queryInfo = dispatchManager.getFullQueryInfo(queryId);
         if (queryInfo.isPresent()) {
-            return Response.ok(queryInfo.get()).build();
+            try {
+                canViewQueryOwnedBy(extractAuthorizedIdentity(servletRequest, accessControl), queryInfo.get().getSession().getUser(), accessControl);
+                return Response.ok(queryInfo.get()).build();
+            }
+            catch (AccessDeniedException e) {
+                throw new ForbiddenException();
+            }
         }
         return Response.status(Status.GONE).build();
     }
 
     @DELETE
     @Path("{queryId}")
-    public void cancelQuery(@PathParam("queryId") QueryId queryId)
+    public void cancelQuery(@PathParam("queryId") QueryId queryId, @Context HttpServletRequest servletRequest)
     {
         requireNonNull(queryId, "queryId is null");
-        dispatchManager.cancelQuery(queryId);
+
+        try {
+            BasicQueryInfo queryInfo = dispatchManager.getQueryInfo(queryId);
+            canKillQueryOwnedBy(extractAuthorizedIdentity(servletRequest, accessControl), queryInfo.getSession().getUser(), accessControl);
+            dispatchManager.cancelQuery(queryId);
+        }
+        catch (AccessDeniedException e) {
+            throw new ForbiddenException();
+        }
+        catch (NoSuchElementException ignored) {
```

Comment:
It wasn't catched before. Why do you need to handle this now? If it was a bug, please extract this to separate commit.